### PR TITLE
Ft toward py3.10

### PR DESCRIPTION
**Pull request by GijsGroote**
Uses a pre-release of casadi compatible with python 3.10 and forces python 3.10 to be allowed only.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -11,9 +11,9 @@ packages = [
   
 
 [tool.poetry.dependencies]
-python = "^3.6.2,<3.10"
+python = "^3.10"
```
Comment (maxspahn) [inline pyproject.toml:14]: We cannot do that as it would end support for python <3.10. 
Rather try it with `^3.6.2,<=3.10`.

File: pyproject.toml [hunk 2]
```diff
@@ -11,9 +11,9 @@ packages = [
...
 numpy = "^1.19.0"
-casadi = "^3.5.4,!=3.5.5.post1,!=3.5.5.post2"
+casadi = "3.5.6.pre2"
```
Comment (maxspahn) [inline pyproject.toml:16]: This casadi library is preciesly the issue with python 3.10. Sometimes it cannot be found directly.

**Review by maxspahn — CHANGES_REQUESTED** [review]
We have tried to upgrade to python3.10 (or at least to support it) before and were blocked by the lack of support of casadi.
See my comments below.
Could you give me the rational behind adding of the function `add_required_key()`? OmegaConf already deals with missing components.

File: .github/workflows/pylint.yml [hunk 3]
```diff
@@ -32,9 +34,14 @@ jobs:
           virtualenvs-create: true
           virtualenvs-in-project: true
           installer-parallel: true
-          version: 1.1.14
+      #----------------------------------------------
+      #     test poetry version 
+      #----------------------------------------------
+      - name: test poetry version
+        run: poetry --version
```
[the remark this answers is not in the document]
Comment (GijsGroote) [reply .github/workflows/pylint.yml:41]: Not really, I did not add it. (but it did came here I think by copying and pasting)

File: .github/workflows/unitTest.yml [hunk 4]
```diff
@@ -6,6 +6,7 @@ name: Testing motion planning scenes
 on: 
   pull_request:
     branches:
+      - develop_python3.10
```
[the remark this answers is not in the document]
Comment (GijsGroote) [reply .github/workflows/unitTest.yml:9]: > Could you give me the rational behind adding of the function `add_required_key()`? OmegaConf already deals with missing components.

Indeed, I only found out in a later stage that OmegaConf checks for missing components. 

The check_completeness and add_required_keys are a centralized result of the add_geometric_keys function which was/is in multiple files. I've moved it to the base class which would then handle all keys.

**Comment by GijsGroote** [comment]
> Could you give me the rational behind adding of the function `add_required_key()`? OmegaConf already deals with missing components.

Indeed, I only found out in a later stage that OmegaConf checks for missing components. 

The check_completeness and add_required_keys are a centralized result of the add_geometric_keys function which was/is in multiple files. I've moved it to the base class which would then handle all keys.

File: pyproject.toml (hunk 2 above)
Comment (GijsGroote) [reply pyproject.toml:16]: it has been updated to "3.5.6.pre2, ^3.5.5,"

**Comment by maxspahn** [comment]
> What is you preference for combining python3.10 with the currently allowed python version?

Could you first give me the rational for why you absolutely need python3.10 for some functionality?
Because having two branches exist in parallel should only be the last resort.

**Comment by maxspahn** [comment]
> There could be some use in these functions (add_required_key and check_completeness), but if there is none. It should be removed. What do you think?

I think we don't need to check functions as omega_conf already does that.

**Comment by GijsGroote** [comment]
check_completeness and add_required_keys are removed.

The main argument to add Python3.10 is to broaden the allowed python versions. That would not be an issue, adding the KW_only (introduced in Py3.10) part breaks backward compatibility , and thus the question becomes should there be 2 separate branches, one for <=py3.9 and one for >=py3.10 only to add KW_ONLY?

I think yes, because:
- it allows (especially ubuntu 22.04 users) to use python3.10  
- it allows for proper inheritance with the use of dataclasses 
- This issue for other features might arise again, I would like to keep the door open for new python features. Better to address this sooner than later.
-  I think allowing new python versions should be more important than mainaining older python versions.

One major drawback which I see is maintaining 2 versions of this repo. I see the 2 versions as a temporary fix, eventually (3,4,5 years later) only >=py3.10 should be allowed. 

**Proper inheritance**
With the refactor often and refactor early principle in mind, I want to use the KW_ONLY parameter, it just feels very wrong to add all the getters, setter, checkers to every class which extends the collisionObstacle class such as sphereObstacle. 
[This](https://medium.com/@aniscampos/python-dataclass-inheritance-finally-686eaf60fbb5)  article shows how data classes can nicely be used.

Dataclasses are not a proper fit for <=py3.9 because it adds code duplication. I suggest for those python versions to keep using [add_shapes](https://github.com/maxspahn/gym_envs_urdf/blob/94e6007353d6e17207d1d694cbfc19acc3b59ad0/urdfenvs/urdfCommon/urdf_env.py#L311) function to add obstacles. From Py3.10 can properly add obstacles.

**Review by maxspahn — APPROVED** [review]

**Pull request merged**